File: service/game/actions/development.py
import uuid

from service.game.actions.base import Command
from service.game.models.development import Development
from service.game.models.map import MapTile


class BuildDevelopmentCommand(Command):
    def execute(self, game_state, player):
        if game_state.phase != "WORK":
            return False

        tile_id = self.payload.get("tile_id")
        target_tile: MapTile = game_state.map_data.get(tile_id)
        if not target_tile or not tile_id:
            return False
        dev_type = target_tile.type
        if not dev_type:
            return False

        existing_dev = (
            target_tile.get("development")
            if isinstance(target_tile, dict)
            else getattr(target_tile, "development", None)
        )
        if existing_dev is not None:
            return False

        build_costs = game_state.development_costs.get(
            dev_type, {}
        ).get("build", {})
        if not build_costs or not self._deduct_resources(
            player, build_costs
        ):
            return False

        dev_id = str(uuid.uuid4())
        new_dev = Development(
            dev_id,
            dev_type,
            player.session_id,
            game_state.rules.MAX_DEVELOPMENT_LEVEL,
            game_state.rules.MAINTENANCE_DAYS,
            game_state.rules.RESOURCE_COSTS,
        )
        game_state.developments[dev_id] = new_dev
        player.developments.append(dev_id)
        target_tile.development = new_dev
        player.add_timeline_event(
            "ACTION_COMPLETED",
            {"action": "BUILD_DEV", "dev_id": dev_id, "type": dev_type},
        )
        player.committed_action = {
            "Action": "Build",
            "Type": dev_type,
            "Tile_Id": tile_id,
        }
        return True


class MaintainDevelopmentCommand(Command):
    def execute(self, game_state, player):
        if game_state.phase != "WORK":
            return False
        dev_id = self.payload.get("dev_id")
        dev = game_state.developments.get(dev_id)
        if not dev or dev.owner != player.session_id:
            return False

        if self._deduct_resources(player, dev.get_maintenance_cost()):
            dev.maintenance()
            player.add_timeline_event(
                "ACTION_COMPLETED",
                {"action": "MAINTAIN_DEV", "dev_id": dev_id},
            )
            return True
        return False


class UpgradeDevelopmentCommand(Command):
    def execute(self, game_state, player):
        if game_state.phase != "WORK":
            return False
        dev_id = self.payload.get("dev_id")
        dev = game_state.developments.get(dev_id)
        if (not dev
                or dev.owner != player.session_id
                or not dev.can_upgrade):
            return False

        if self._deduct_resources(player, dev.get_upgrade_cost()):
            dev.upgrade()
            player.add_timeline_event(
                "ACTION_COMPLETED",
                {"action": "UPGRADE_DEV", "dev_id": dev_id},
            )
            return True
        return False

File: service/src/game/actions/development.ts
import type {
  DevelopmentDTO,
  GameStateDTO,
  PlayerDTO,
} from "@takes-a-village/shared";

interface DevelopmentPlayer {
  readonly id: string;
  readonly resources: PlayerDTO["resources"];
  finishedPhase: boolean;
}

interface DevelopmentGame {
  readonly phase: GameStateDTO["phase"];
  readonly developments: Map<string, DevelopmentDTO>;
  readonly rules: {
    readonly MAINTENANCE_DAYS: number;
    readonly MAX_DEVELOPMENT_LEVEL: number;
  };
}

export function handleDevelopmentAction(
  game: DevelopmentGame,
  player: DevelopmentPlayer,
  command: unknown,
  payload: Record<string, unknown>,
): boolean {
  if (game.phase !== "WORK" || typeof payload.dev_id !== "string") return false;
  const development = game.developments.get(payload.dev_id);
  if (!development || development.owner_id !== player.id) return false;
  if (command === "MAINTAIN_DEV") {
    if (!deduct(player.resources, development.maintenance_cost)) return false;
    development.maintenance_days = game.rules.MAINTENANCE_DAYS;
    player.finishedPhase = true;
    return true;
  }
  if (command === "UPGRADE_DEV") {
    if (!development.can_upgrade || !deduct(player.resources, development.upgrade_cost)) return false;
    development.level = Math.min(development.level + 1, game.rules.MAX_DEVELOPMENT_LEVEL);
    development.maintenance_days = game.rules.MAINTENANCE_DAYS;
    development.can_upgrade = development.level < game.rules.MAX_DEVELOPMENT_LEVEL;
    development.maintenance_cost = maintenanceCost(development);
    development.upgrade_cost = upgradeCost(development);
    player.finishedPhase = true;
    return true;
  }
  return false;
}

export function degradeDevelopments(game: DevelopmentGame): void {
  for (const development of game.developments.values()) {
    development.maintenance_days -= 1;
    if (development.maintenance_days >= 0) continue;
    if (development.level > 1) {
      development.level -= 1;
      development.maintenance_days += game.rules.MAINTENANCE_DAYS;
    } else {
      development.level = 1;
      development.maintenance_days = 1;
    }
    development.can_upgrade = development.level < game.rules.MAX_DEVELOPMENT_LEVEL;
    development.maintenance_cost = maintenanceCost(development);
    development.upgrade_cost = upgradeCost(development);
  }
}

function maintenanceCost(development: DevelopmentDTO): Record<string, number> {
  if (development.type === "Farm") return { wood: development.level, iron: Math.max(development.level - 1, 0) };
  if (development.type === "Woods") return { food: development.level, iron: Math.max(development.level - 1, 0) };
  const amount = development.level * 2 + 1;
  return { food: amount, wood: amount };
}

function upgradeCost(development: DevelopmentDTO): Record<string, number> {
  if (development.type === "Farm") return { wood: development.level * 2 + 1, iron: development.level };
  if (development.type === "Woods") return { food: development.level * 2 + 1, iron: development.level };
  return { food: development.level, wood: development.level, iron: development.level * 2 + 1 };
}

function deduct(resources: PlayerDTO["resources"], costs: Record<string, number>): boolean {
  for (const [resource, amount] of Object.entries(costs)) {
    const key = resource as keyof PlayerDTO["resources"];
    if (resources[key] < amount) return false;
  }
  for (const [resource, amount] of Object.entries(costs)) {
    const key = resource as keyof PlayerDTO["resources"];
    resources[key] -= amount;
  }
  return true;
}
